### PR TITLE
fix(nlpgo): inline saved datasets + parallelize eval + sklearn-parity train/test split (3.2.0 prod fix)

### PR DESCRIPTION
**Pull request by rogeriochaves**
## Summary

3.2.0 prod regression: a customer hit `entry node has no inline dataset (remote datasets not yet supported on Go path)` on a saved-dataset Evaluate run. Three S1/S2 fixes from the rchaves 2026-04-30 callout, plus parity sweeps and CodeRabbit fixes.

### S1 — Saved (database) datasets work on Go path again

Studio's `loadDatasets()` short-circuited the database-dataset branch when `entry_selection === "all"` (which is what execute_evaluation forces), forwarding a `dataset_id`-only payload. Python's SDK fetched via `langwatch.dataset.get_dataset(id)` — Go had no equivalent, so the engine rejected with the actionable-but-misleading "remote datasets not supported" error.

**Fix:** always inline server-side via `getFullDataset()`. Both engines see the uniform inline shape; Python's path is unchanged semantically (it took the inline branch already when given inline data). Engine guard error rewritten to point operators at the right TS helper.

### S2.1 — Eval rows now run in parallel (10 workers, Python parity)

Python's DSPy Evaluate uses `num_threads=10` (`langwatch_nlp/studio/execute/execute_evaluation.py:144`). Pre-fix the Go path iterated sequentially — ~10x slower for a 26-row dataset. Worker pool capped at `min(10, total)`.

### S2.2 — train/test split shuffle parity with sklearn

Python passes `shuffle=(seed >= 0), random_state=seed` to `sklearn.train_test_split`. Pre-fix `splitTrainTest` ignored the seed and just took an in-order slice. Now Fisher-Yates with seeded RNG when `seed >= 0`, in-order fallback otherwise. Also honors explicit `testSize` (was discarded as `_ = testSize`).

### S2.3 — `type: "custom"` nodes route through `agentblock.WorkflowRunner`

Studio's `NodeSelectionPanel.tsx:168` actively writes `type: "custom"` for saved sub-workflow drag-drops, with typed `data.workflow_id` / `data.version_id`. Pre-fix the planner blanket-rejected this kind as "retired" with a misleading replace-the-node hint, even though the executor (`agentblock.WorkflowRunner`) was already wired and tested for `agent_type: workflow`.

**Fix:** new `runCustom` dispatcher reads typed Component fields (with parameters[] fallback for legacy shapes) and reuses the same WorkflowRunner. Mirrors Python's `CustomNode.forward` — same wire (`POST /api/workflows/<id>[/<v>]/run` with `do_not_trace=true`), same auth header (X-Auth-Token from `workflow.api_key`).

### S2.4 — Full Liquid templates in LLM Call instructions (in flight)

Python's path uses `python-liquid` (full Liquid: `{% for %}` / `{% if %}` / pipe filters / dotted paths / native list iteration). Go's `template.Render` was partial-Liquid only. Customer LLM Call instructions with control-flow tags rendered the literal `{% for %}` text on the Go path. Replacing with `osteele/liquid` for the LLM-instructions surface.

### S2.5 — `/v1beta/*` Gemini-native passthrough wired through dispatcher.Passthrough

Pre-fix returned a typed 501 `passthrough_not_implemented` because the shim adapter didn't carry HTTP-shape fields (method, raw query, forwarded headers). External callers (gemini-cli, Google GenAI SDK direct) were blocked.

**Fix:**
- `DispatcherShim` extended with `Passthrough` + `PassthroughStream` methods + `PassthroughDispatchRequest` type
- `playgroundDispatcherShim` in cmd/root.go maps to `dispatcher.Passthrough`/`PassthroughStream` with `domain.PassthroughRequest`
- New `filterPassthroughHeaders` strips `x-litellm-*` (already in Credential), `Authorization`/`X-Auth-Token`, hop-by-hop, Host, Content-Length, AND headers nominated by `Connection: X, Y` per RFC 7230 §6.1
- `classifyPath` bug fix: was trimming `/v1` even from `/v1beta`, leaving `beta/models/...`
- `isPassthroughStreamPath` promotes `isStream` for `:streamGenerateContent` / `:streamRawPredict` actions (Gemini-native streaming uses URL suffix, not body.stream flag)

### A1 — `do_not_trace` + `workflow.enable_tracing=false` parity

Python's `execute_flow.py:53` honors envelope-level `event.do_not_trace` AND `workflow.enable_tracing=false` to suppress trace emission via `optional_langwatch_trace(do_not_trace=...)`. Go ignored both. Bug pre-fix: any sub-workflow dispatch (`agent_type=workflow` OR `type:custom` — both inject `do_not_trace=true` on the /run body) caused Go to emit a duplicate trace span against the parent.

**Fix:** decode envelope-level `do_not_trace` + peek `workflow.enable_tracing` → OR into `req.DoNotTrace`. `startStudioSpan` returns the no-op span-from-context when set, so engine descendants stay silent too.

### Saved-dataset Evaluate populates Studio's bottom Evaluations panel

The load-bearing assertion for S1 — Studio → Evaluate → Run Evaluation on a workflow whose entry node uses a saved dataset (no inline records). Pre-fix the Evaluate flow crashed with the no-inline-dataset error; post-fix Run #3 appears in the bottom panel within ~1s of clicking Run Evaluation, with row 1 → row N rendered live as `evaluation_state_change progress` events fire.

### Trace-parity scenarios re-verified on the post-fix stack

Single-component (`execute_component`): `nlpgo.studio.execute_stream → execute_component → LLM openai/gpt-5.2` (14 tok / $0.00009 / 1.8s, langwatch.origin=workflow, gen_ai.provider.name=openai)

Whole-flow (`execute_flow`): full SSE chain with both `execution_state_change` running→success bookends, no `Waiting for runtime` stuck state. Trace tree: `execute_stream → execute_component (entry) → execute_component (llm) → LLM openai/gpt-5.2 → execute_component (end)` (21 tok / $0.00020 / 1.2s)

Experiment-detail page (post-S1) — Run #2 actually populates `/experiments/<id>` with question/Output/duration

## Tests

- `langwatch/src/optimization_studio/server/__tests__/loadDatasets.test.ts` — saved-DB inline on execute_evaluation/full + inline preservation + execute_component dataset-strip
- `TestSelectEvaluationEntries_TrainTestSplitShuffleByPositiveSeed` / `NegativeSeedKeepsOrder` / `TestSizeRespected` — sklearn-parity pins
- `TestSelectEvaluationEntries_NonInlineDatasetReturnsActionableError` — guard error wording pin
- `TestEvaluationWorkflow_PostsBatchResultsToLangWatch` (integration) — saved-dataset eval populates experiment-runs page
- `TestPlan_AcceptsCustomKind` + `TestPattern016_CustomNodeKindRoutesToWorkflowRunner` + `TestPattern010_CustomNodeMissingWorkflowIDFailsActionably` — type:custom routing
- `TestPlaygroundProxy_PassthroughForwardsHTTPShapeToDispatcher` — /v1beta wire shape (HTTPMethod/HTTPPath/HTTPRawQuery/HTTPHeaders, Authorization stripped, Credential reaches dispatcher)
- `TestPlaygroundProxy_PassthroughStreamPathDetectsActionSuffix` — `:streamGenerateContent` / `:streamRawPredict` promote `isStream`; unknown actions stay sync
- `TestFilterPassthroughHeaders_StripsConnectionNominatedHeaders` — RFC 7230 §6.1 dynamic denylist
- `TestDecodeStudioClientEvent_DoNotTrace_FromEnvelope` / `FromEnableTracingFalse` / `DefaultsFalse` + `TestPeekWorkflowEnableTracing` — A1 audit pins

## Test plan

- [x] `pnpm test:unit src/optimization_studio/server/__tests__/loadDatasets.test.ts`
- [x] `go test ./services/nlpgo/...` — full nlpgo suite green
- [x] Browser dogfood: Studio → Evaluate on a saved-dataset workflow → Studio's bottom Evaluations panel populates with the new run (screenshot above)
- [x] Browser dogfood: trace-parity scenarios 1+2 re-verified on the post-fix stack
- [x] CodeRabbit review iteration — comments addressed inline

### CODE REVIEW

File: services/nlpgo/adapters/httpapi/playground_proxy.go [hunk 1]
```diff
@@ -156,6 +148,21 @@ func playgroundProxyDispatch(proxy PlaygroundProxy) http.HandlerFunc {
...
+		// query, forwarded headers) so the dispatcher can reconstruct the
+		// upstream call. Auth headers (x-litellm-* — already parsed into
+		// Credential above; Authorization/X-Auth-Token — never expected
+		// from a /v1beta caller in our deployment shape) are stripped so
+		// they aren't double-applied or leaked to the upstream provider.
+		if reqType == domain.RequestTypePassthrough {
+			req.HTTPMethod = r.Method
+			req.HTTPRawQuery = r.URL.RawQuery
+			req.HTTPHeaders = filterPassthroughHeaders(r.Header)
+		}
```
[the remark this answers is not in the document]
Comment (rogeriochaves) [reply services/nlpgo/adapters/httpapi/playground_proxy.go:175]: Addressed in `3b552e337` — new `isPassthroughStreamPath` helper recognizes `:streamGenerateContent` / `:streamRawPredict` action suffixes and promotes `isStream` for them, since Gemini-native streaming signals via the URL action on `:` rather than via OpenAI's body.stream flag. 6-case unit test pin (`TestPlaygroundProxy_PassthroughStreamPathDetectsActionSuffix`) covers sync/stream/missing/unknown paths. Falls back to non-stream for unknown actions so a typo surfaces as a regular sync 4xx instead of an opened SSE stream that never closes.

File: services/nlpgo/adapters/httpapi/playground_proxy.go [hunk 2]
```diff
@@ -421,3 +501,39 @@ func (a *shimAdapter) DispatchStream(ctx context.Context, req playgroundProxyReq
...
+			continue
+		}
+		// Multi-value headers are joined with comma per RFC 7230 §3.2.2 —
+		// matches what http.Client does on the way out anyway.
+		if len(vs) > 0 {
+			out[k] = strings.Join(vs, ", ")
+		}
+	}
+	return out
+}
```
[the remark this answers is not in the document]
Comment (rogeriochaves) [reply services/nlpgo/adapters/httpapi/playground_proxy.go:588]: Addressed in `3b552e337` — `filterPassthroughHeaders` now parses the `Connection: X, Y` token list and adds each token to a per-call dynamic skip set BEFORE copying headers. So `Connection: Foo` plus `Foo: bar` correctly drops both upstream. Test pin `TestFilterPassthroughHeaders_StripsConnectionNominatedHeaders` asserts `X-Hop-By-Hop` + `Trailer-Test` (both nominated by Connection) are stripped while `Accept` + `X-LangWatch-Trace-Id` pass through, plus the canonical denylist (Connection itself, Authorization, x-litellm-*) still fires.

**Review by sergioestebance — APPROVED** [review]
lgtm
